File: SearchFuncs.py
from math import ceil, floor
from string import ascii_letters
keywords = {
    '+',
    '-',
    '*',
    '/',
    '!',
    '^',
    's','t','c','r','l','R','S','%'
}

non_keywords = (set(ascii_letters).difference(keywords)) | {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'}


def searchChar(expr, char):
    parenthesisCounter = 0
    absCounter = 0
    for i in range(len(expr) - 1, -1, -1):
        if expr[i] == ")":
            parenthesisCounter += 1
        elif expr[i] == "(":
            parenthesisCounter -= 1
        elif expr[i] == "|":
            if isClosingAbs(expr, i):
                absCounter -= 1
            else:
                absCounter += 1
        elif expr[i] == char and parenthesisCounter == 0 and absCounter == 0 and expr[i - 1] not in (
        "*", "/", "^", "-", "+", "√"):
            return i
    return -1


def searchSplittingComma(expr, idxLog):
    logCounter = 0
    parCounter = 0
    for i in range(idxLog + 2, len(expr)):
        if expr[i] == 'l' or expr[i] == 'S':
            logCounter += 1
        elif expr[i] == '(':
            parCounter += 1
        elif expr[i] == ')':
            if parCounter == 1:
                logCounter -= 1
            parCounter -= 1
        elif expr[i] == ',' and logCounter == 0:
            return i
    return -1

def searchSigma(expr):
    parenthesisCounter = 0
    afterMiddle = False
    absCounter = 0
    for i in range(len(expr) - 1, -1, -1):
        if absCounter >= expr.count("|") / 2:
            afterMiddle = True
        if expr[i] == ")":
            parenthesisCounter += 1
        elif expr[i] == "(":
            parenthesisCounter -= 1
        elif expr[i] == "|":
            if afterMiddle:
                absCounter -= 1
            else:
                absCounter += 1
        elif expr[i] == "S" and parenthesisCounter == 0 and absCounter == 0 and expr[i - 1] not in (
        "*", "/", "^", "-", "+", "√"):
            break
    else:
        return -1
    return splitSigma(expr, i)


def splitSigma(expr, idxSum):
    sigmaCounter = 0
    Is = []
    parCounter = 0
    for i in range(idxSum+2, len(expr)-1):
        if len(Is) == 3:
            break
        if expr[i] in ("S", 'l'):
            sigmaCounter += 1
        elif expr[i] == '(':
            parCounter += 1
        elif expr[i] == ')':
            if parCounter == 1:
                sigmaCounter -= 1
            parCounter -= 1
        elif expr[i] == ',' and sigmaCounter == 0:
            Is.append(i)
    a, b = Is

    start, end, func = expr[2:a], expr[a+1:b], expr[b+1:-1]
    var, start = start.split("=")
    return var, start, end, func


def isClosingAbs(expr, i):
    if i == 0:
        return False
    if i == len(expr)-1:
        return True

    left, right = expr[i-1], expr[i+1]
    if right == '|':
        return isClosingAbs(expr, i+1)
    if left == '|':
        return isClosingAbs(expr, i-1)
    if left in non_keywords and right in non_keywords:
        return False
    if right in (keywords.difference({'-', '+'})):
        return True
    if left.isnumeric():
        return True
    return False


def searchDivMul(expr, c):
    parenthesisCounter = 0
    for i in range(len(expr) - 1, -1, -1):
        if expr[i] == ")":
            parenthesisCounter += 1
        elif expr[i] == "(":
            parenthesisCounter -= 1
        elif expr[i] == c and parenthesisCounter == 0:
            return i
    return -1

def searchPlus(expr):
    parenthesisCounter = 0
    for i in range(len(expr)-1, -1, -1):
        if expr[i] == ")":
            parenthesisCounter += 1
        elif expr[i] == "(":
            parenthesisCounter -= 1
        elif expr[i] == '+' and parenthesisCounter == 0:
            if expr[i-1] in ("*", "/", "^", "-", "+", "√", "!"):
                continue
            return i
    return -1

def searchPow(expr):
    parenthesisCounter = 0
    for i, char in enumerate(expr):
        if char == "(":
            parenthesisCounter += 1
        elif char == ")":
            parenthesisCounter -= 1
        elif char == "^" and parenthesisCounter == 0:
            return i
    return -1

def searchMinus(expr):
    parenthesisCounter = 0
    for i in range(len(expr)-1, -1, -1):
        if expr[i] == ")":
            parenthesisCounter += 1
        elif expr[i] == "(":
            parenthesisCounter -= 1
        elif expr[i] == '-' and parenthesisCounter == 0:
            if expr[i-1] in ("*", "/", "^", "-", "+", "√", "!"):
                continue
            return i
    return -1

File: Calculator/SearchFuncs.cpp
#include "SearchFuncs.h"
#include "Arduino.h"
#include "Operators.h"

bool isIn(char c, char arr[]){
  for (byte i = 0; i < strlen(arr); i++){
    if (arr[i] == c)
      return true;
  }
  return false;
}

byte count(String s, char c){
  byte n = 0;
  for(byte i = 0; i < s.length(); i++){
    if (s[i] == c)
      n++;
  }
  return n;
}

bool isClosingAbs(String expr, unsigned i){
    if (i == 0)
        return false;
    if (i == expr.length()-1)
        return true;

    char left = expr[i-1];
    char right = expr[i+1];
    
    if (right == '|')
      return isClosingAbs(expr, i+1);
    if (left == '|')
      return isClosingAbs(expr, i-1);

    if(isIn(left, keywords)) // for sure
      return false;

    // for sure
    if (isIn(left, nonKeywords)) // (canConvertToInt((String)left))
      return true;
    if (isIn(left, nonKeywords) && isIn(right, nonKeywords))
      return false;
    if (isIn(right, keywords) && right != '-' && right != '+')
      return true;

    return false;
}

int searchSplittingComma(String expr, int idxLog){
  byte logCounter = 0;
  byte parCounter = 0;
  for (byte i = idxLog+2; i < expr.length(); i++){
    if (expr[i] == logC || expr[i] == sumC)
      logCounter++;
    else if(expr[i] == '(')
      parCounter++;
    else if(expr[i] == ')'){
      if (parCounter == 1)
        logCounter--;
      parCounter--;
    }
    else if (expr[i] == ',' && logCounter == 0)
      return i;
  }
  return -1;
}

String splitSigma(String expr, int idxSum){
  byte sigmaCounter = 0;
  int a = -1;
  int b = -1;
  int parCounter = 0;
  for (byte i = idxSum+2; i < expr.length()-1; i++){
    if (b != -1)
      break;
    if (expr[i] == sumC || expr[i] == logC){
      sigmaCounter += 1;
    }
    else if (expr[i] == '('){
      parCounter += 1;
    }
    else if (expr[i] == ')'){
      if (parCounter == 1){
          sigmaCounter -= 1;
      }
      parCounter -= 1;
    }
    else if (expr[i] == ',' && sigmaCounter == 0){
      if(a == -1)
        a = i;
      else
        b = i;
    } 
  }

  // the x=3 part
  String varVal = expr.substring(2, a);
  String End = expr.substring(a+1, b);
  String func = expr.substring(b+1, expr.length()-1);
  
  // var
  String var = varVal.substring(0, varVal.indexOf('='));
  String Val = varVal.substring(varVal.indexOf('=')+1);

  //_var = var;
  //_start = Val;
  //_end = End;
  //_func = func;
  return var + " " + Val + " " + End + " " + func;
}

int searchSigma(String expr){
  byte parCount = 0;
  byte absCounter = 0;
  
  for(byte i = 0; i < expr.length(); i++){
    if (expr[i] == ')')
      parCount++;
    else if(expr[i] == '(')
      parCount--;
    else if (expr[i] == '|'){
      if (isClosingAbs(expr, i))
        absCounter -= 1;
      else
        absCounter += 1;
    }
    else if(expr[i] == sumC && parCount == 0 && absCounter == 0 &&
        ((expr[i-1] != '+' &&
        expr[i-1] != '-' &&
        expr[i-1] != '*' &&
        expr[i-1] != '/' &&
        expr[i-1] != '^' ) ||
        i == 0))
      return i;
  }
 return -1;
}

int searchChar(String expr, char c){
  byte parCount = 0;
  byte absCounter = 0;
  
  for(int i = expr.length()-1; i >= 0; i--){
    if (expr[i] == ')')
      parCount++;
    else if(expr[i] == '(')
      parCount--;
    else if (expr[i] == '|'){
      if (isClosingAbs(expr, i))
        absCounter -= 1;
      else
        absCounter += 1;
    }
    else if(expr[i] == c && parCount == 0 && absCounter == 0 &&
        (expr[i-1] != '+' &&
        expr[i-1] != '-' &&
        expr[i-1] != '*' &&
        expr[i-1] != '/' && (
        expr[i-1] != '^' ||
        i == 0)))
      return i;
  }
 return -1;
}
